sstables/compaction.cc

```diff
@@ -1230,38 +1269,29 @@ class resharding_compaction final : public compaction {
         throw std::runtime_error("create_new_sstable called from resharding. Shouldn't have happened!!");
     }
 
-    sstable_writer* select_sstable_writer(const dht::decorated_key& dk) override {
-        _shard = dht::shard_of(*_schema, dk.token());
-        auto& sst = _output_sstables[_shard].first;
-        auto& writer = _output_sstables[_shard].second;
-
-        if (!writer) {
-            sst = _sstable_creator(_shard);
-            setup_new_sstable(sst);
+    compaction_writer create_compaction_writer(const dht::decorated_key& dk) override {
+        auto shard = dht::shard_of(*_schema, dk.token());
+        auto sst = _sstable_creator(shard);
+        setup_new_sstable(sst);
 
-            sstable_writer_config cfg = _cf.get_sstables_manager().configure_writer();
-            cfg.max_sstable_size = _max_sstable_size;
-            // sstables generated for a given shard will share the same run identifier.
-            cfg.run_identifier = _run_identifiers.at(_shard);
-            auto&& priority = service::get_local_compaction_priority();
-            writer.emplace(sst->get_writer(*_schema, partitions_per_sstable(_shard), cfg, get_encoding_stats(), priority, _shard));
-        }
-        return &*writer;
+        sstable_writer_config cfg = _cf.get_sstables_manager().configure_writer();
+        cfg.max_sstable_size = _max_sstable_size;
+        // sstables generated for a given shard will share the same run identifier.
+        cfg.run_identifier = _run_identifiers.at(shard);
+        auto&& priority = service::get_local_compaction_priority();
+        return compaction_writer{sst->get_writer(*_schema, partitions_per_sstable(shard), cfg, get_encoding_stats(), priority, shard), sst};
     }
 
-    void stop_sstable_writer() override {
-        auto& sst = _output_sstables[_shard].first;
-        auto& writer = _output_sstables[_shard].second;
+    void on_new_partition() override {}
 
-        finish_new_sstable(writer, sst);
+    void stop_sstable_writer(compaction_writer* writer) override {
+        if (writer) {
+            finish_new_sstable(writer);
+        }
     }
 
-    void finish_sstable_writer() override {
-        for (auto& p : _output_sstables) {
-            if (p.second) {
-                finish_new_sstable(p.second, p.first);
-            }
-        }
+    void finish_sstable_writer(compaction_writer* writer) override {
+        finish_new_sstable(writer);
```

Comment:
finish_new_sstable() unconditionally dereferences writer, so it shouldn't be called unconditionally here given that compaction writer can be NULL. Like in regular_compaction, let's call stop_sstable_writer() instead of directly calling finish_new_sstable().

Reply:
Good one, thanks